idaes/apps/grid_integration/coordinator.py

```diff
@@ -0,0 +1,879 @@
+# Design of Advanced Energy Systems (IDAES), and is copyright (c) 2018-2021
+# by the software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia University
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+from pyomo.common.dependencies import attempt_import
+
+prescient, prescient_avail = attempt_import("prescient")
+if prescient_avail:
+    import prescient.plugins as pplugins
+    from prescient.simulator.config import PrescientConfig
+from pyomo.common.config import ConfigDict, ConfigValue
+import pyomo.environ as pyo
+
+
+class DoubleLoopCoordinator:
+
+    """
+    Coordinate Prescient, tracker and bidder.
+    """
+
+    def __init__(self, bidder, tracker, projection_tracker, self_schedule=False):
+
+        """
+        Initializes the DoubleLoopCoordinator object and registers functionalities
+        in Prescient's plugin system.
+
+        Arguments:
+            bidder: an initialized bidder object
+
+            tracker: an initialized bidder object
+
+            projection_tracker: an initialized bidder object, this object is for
+                                projecting the system states and updating bidder
+                                model
+
+            self_schedule: whether the resource is participating through self-scheduling
+
+        Returns:
+            None
+        """
+
+        self.bidder = bidder
+        self.tracker = tracker
+        self.projection_tracker = projection_tracker
+        self.self_schedule = self_schedule
+
+    def register_plugins(
+        self,
+        context,
+        options,
+        plugin_config,
+    ):
+
+        """
+        Register functionalities in Prescient's plugin system.
+
+        Arguments:
+            context: Prescient plugin PluginRegistrationContext
+
+            options: Prescient options
+
+            plugin_config: Prescient plugin config
+
+        Returns:
+            None
+        """
+
+        self.plugin_config = plugin_config
+
+        self._register_initialization_callbacks(context, options, plugin_config)
+        self._register_before_ruc_solve_callbacks(context, options, plugin_config)
+        self._register_before_operations_solve_callbacks(
+            context, options, plugin_config
+        )
+        self._register_after_operations_callbacks(context, options, plugin_config)
+        self._register_update_operations_stats_callbacks(
+            context, options, plugin_config
+        )
+        self._register_after_ruc_activation_callbacks(context, options, plugin_config)
+        self._register_finalization_callbacks(context, options, plugin_config)
```

Comment:
What is the value of these 7 1-line functions?  Why don't you just inline them here?

Reply:
These functions don't return anything. Should I replace these with the actual implementation of the functions?

Reply:
> These functions don't return anything. Should I replace these with the actual implementation of the functions?

Yeah, I think that is cleaner as well, but probably not critical for merging.

Reply:
Thanks!